### PR TITLE
Wrap operations as `TensorMap` and  `TensorBlock` methods

### LINKED ISSUE #40: Better error message when trying to assign to block.values directly
It could be useful to manipulate the descriptor values of a block after they have been calculated. However, expressions like 

```python
block.values *= -1
```

are not allowed an lead to `AttributeError: can't set attribute`. There is a workaround like 

```python
for r in block.values:
    r *= -1
```

which works fine, but in my view the first option should also work.

### PR DESCRIPTION
**Pull request by PicoCentauri**
Adding basic methods for math and manipulation operations on a `Tensormap` and `TensorBlock` class.
This will give users a lot more usability when working with `Tensormaps` and `TensorBlocks`.

The implementation here does not solve the discussion of #118. It only wraps the already implemented operations as class methods.

My approach for the imports is a bit hacky... I was not able to put the imports at the top of the file because of [circular import issues](https://stackoverflow.com/questions/7336802/how-to-avoid-circular-imports-in-python). We require the TensorBlock and TensorMap class in the operations but also the operations in the TensorMap and TensorBlock class. If somebody knows how to solve them, I am happy to learn.

We could think of adding the same basic math operations also for blocks. But, we can do this in an upcoming PR.

# TODO
- [x] Tests
- [x] Check docs rendering

----
:books: Documentation preview :books:: https://equistore--141.org.readthedocs.build/en/141/

### CODE REVIEW

File: python/src/equistore/tensor.py [hunk 1]
```diff
@@ -105,6 +105,84 @@ def __getitem__(self, *args) -> TensorBlock:
...
+
+        return pow(self, other)
+
+    def __iadd__(self, other):
+        from equistore.operations import add
+
+        return add(self, other)
+
+    def __isub__(self, other):
+        from equistore.operations import add
```
Comment (DavideTisi) [inline python/src/equistore/tensor.py:154]: subtract

File: python/src/equistore/tensor.py [hunk 2]
```diff
@@ -394,6 +460,146 @@ def property_names(self) -> List[str]:
...
+
+        :param samples_names:
+            names of samples to perform the standart deviation over
+        """
+        return self._reduce_over_samples(samples_names, "std")
+
+    def variance_over_samples(self, samples_names: List[str]) -> "TensorMap":
+        r"""Variance over the indices of the given ``samples_names``.
+
+        For addiotional details refer to the reference page of
```
Comment (DavideTisi) [inline python/src/equistore/tensor.py:537]: addiotional

Comment (PicoCentauri) [reply python/src/equistore/tensor.py:537]: Thanks @DavideTisi for always pointing out my typos 🤣

**Review by DavideTisi — APPROVED** [review]

File: python/src/equistore/tensor.py [hunk 3]
```diff
@@ -105,6 +105,72 @@ def __getitem__(self, *args) -> TensorBlock:
...
+        from equistore.operations import divide
+
+        return divide(self, other)
+
+    def __pow__(self, other):
+        from equistore.operations import pow
+
+        return pow(self, other)
+
+    def __iadd__(self, other):
```
Comment (Luthaf) [inline python/src/equistore/tensor.py:148]: I would not do the in-place operations if they are not actually in-place

File: python/src/equistore/tensor.py [hunk 4]
```diff
@@ -394,6 +460,146 @@ def property_names(self) -> List[str]:
...
+        :py:class:`TensorBlocks` corresponding to the same key must have the same
+        ``properties``. The resulting :py:class:`TensorBlocks` of the dot product of
+        two :py:class:`TensorBlocks` has ``result_block.values = block1.values @
+        block2.values.T``
+
+        :param other:
+            :py:class:`TensorMap` to multiply
+
+        :return: dot product
+        """
```
Comment (Luthaf) [inline python/src/equistore/tensor.py:489]: I don't love that we have to duplicate the documentation here

File: python/src/equistore/tensor.py [hunk 5]
```diff
@@ -394,6 +460,146 @@ def property_names(self) -> List[str]:
...
+        For additional details refer to the reference page of
+        the :ref:`python-api-operations-sample-reduction` functions.
+
+        :param samples_names:
+            names of samples to sum over
+        """
+        return self._reduce_over_samples(samples_names, "sum")
+
+    def mean_over_samples(self, samples_names: List[str]) -> "TensorMap":
+        """Aaverage over the indices of the given ``samples_names`` .
```
Comment (DavideTisi) [inline python/src/equistore/tensor.py:513]: Aaverage

File: python/src/equistore/tensor.py (hunk 3 above)
Comment (PicoCentauri) [reply python/src/equistore/tensor.py:148]: Yes, I agree.

File: python/src/equistore/tensor.py (hunk 4 above)
Comment (PicoCentauri) [reply python/src/equistore/tensor.py:489]: We can do it numpy style and briefly state.

```
"""Computes the dot product with another :py:class:`TensorMap`.

Refer to :py:func:`equistore.dot` for full documentation.
"""
```

**Review by DavideTisi — APPROVED** [review]

**Comment by Luthaf** [comment]
While I agree that this makes sense for the arithmetic operations (`__add__` and friends), I'm less sure about the other operations (`sum_samples`, …)

One issue is that while the core code is reasonably stable and could be made into a 0.1 release, the operations are still moving a lot. We could make a release soon by not including the operations in the set of stable functions (e.g. putting them in `equistore.unstable.XXX` until we are confident on the API), but exposing all of them as member functions of TensorMap does not allow us to select which ones we want to make stable.

**Comment by PicoCentauri** [comment]
I see your point. I would be fine if for now we only add the arithmetic operations.

Once we agree on the operations I am definitely in favor of adding them as well.

**Comment by PicoCentauri** [comment]
I removed the inplace operations. For me this looks still like a useful addition. Thoughts?

File: python/src/equistore/operations/reduce_over_samples.py [hunk 6]
```diff
@@ -306,7 +309,9 @@ def mean_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
 
 
 def std_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
-    r"""Create a new :py:class:`TensorMap` with the same keys as
+    r"""Standard deviation over the indices of the given ``samples_names``.
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:312]:
```suggestion
    r"""Compute the standard deviation of a :py:class:`TensorMap`, grouping samples according to ``samples_names``.
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 7]
```diff
@@ -335,7 +340,9 @@ def std_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
 
 
 def variance_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
-    r"""Create a new :py:class:`TensorMap` with the same keys as
+    r"""Variance over the indices of the given ``samples_names``.
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:343]:
```suggestion
    r"""Compute the variance of a :py:class:`TensorMap`, grouping samples according to ``samples_names``.
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 8]
```diff
@@ -335,7 +340,9 @@ def std_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
...
+
+    Create a new :py:class:`TensorMap` with the same keys as
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:345]:
```suggestion
    This function creates a new :py:class:`TensorMap` with the same keys as
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 9]
```diff
@@ -306,7 +309,9 @@ def mean_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
 
 
 def std_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
-    r"""Create a new :py:class:`TensorMap` with the same keys as
+    r"""Standard deviation over the indices of the given ``samples_names``.
+
+    Create a new :py:class:`TensorMap` with the same keys as
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:314]:
```suggestion
    This function creates a new :py:class:`TensorMap` with the same keys as
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 10]
```diff
@@ -284,7 +285,9 @@ def sum_over_samples(
 
 
 def mean_over_samples(tensor: TensorMap, samples_names: List[str]) -> TensorMap:
-    """Create a new :py:class:`TensorMap` with the same keys as
+    """Mean over the indices of the given ``samples_names``.
+
+    Create a new :py:class:`TensorMap` with the same keys as
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:290]:
```suggestion
    This function creates a new :py:class:`TensorMap` with the same keys as
```

File: python/src/equistore/operations/reduce_over_samples.py [hunk 11]
```diff
@@ -219,7 +219,8 @@ def _reduce_over_samples(
 def sum_over_samples(
     tensor: TensorMap, samples_names: Union[List[str], str]
 ) -> TensorMap:
-    """
+    """Sum over the indices of the given ``samples_names``.
+
     Create a new :py:class:`TensorMap` with the same keys as as the input
```
Comment (Luthaf) [inline python/src/equistore/operations/reduce_over_samples.py:224]:
```suggestion
    This function creates a new :py:class:`TensorMap` with the same keys as as the input
```

File: python/src/equistore/operations/remove_gradients.py [hunk 12]
```diff
@@ -7,14 +7,13 @@ def remove_gradients(
     tensor: TensorMap,
     remove: Optional[List[str]] = None,
 ) -> TensorMap:
-    """
-    Create a new :py:class:`TensorMap` without some or
-    all of the gradients from ``tensor``.
+    """Remove some or all of the gradients.
```
Comment (Luthaf) [inline python/src/equistore/operations/remove_gradients.py:10]:
```suggestion
    """Remove some or all of the gradients from a :py:class:`TensorMap`.
```

File: python/src/equistore/operations/slice.py [hunk 13]
```diff
@@ -12,12 +12,11 @@ def slice(
     samples: Optional[Labels] = None,
     properties: Optional[Labels] = None,
 ) -> TensorMap:
-    """
-    Slices an input :py:class:`TensorMap` along the samples and/or properties
-    dimension(s). ``samples`` and ``properties`` are
-    :py:class:`Labels` objects that specify the samples/properties
-    (respectively) names and indices that should be sliced, i.e. kept in the
-    output tensor.
+    """Slice along the samples and/or properties dimension(s).
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:15]:
```suggestion
    """Slice a :py:class:`TensorMap` along the samples and/or properties dimension(s).
```

File: python/src/equistore/operations/split.py [hunk 14]
```diff
@@ -11,9 +11,9 @@ def split(
     axis: str,
     grouped_idxs: List[Labels],
 ) -> List[TensorMap]:
-    """
-    Splits an input :py:class:`TensorMap` into mutliple :py:class:`TensorMap`
-    objects based on some specified groups of indices, along either the
+    """Split into mutliple :py:class:`TensorMap`.
```
Comment (Luthaf) [inline python/src/equistore/operations/split.py:14]:
```suggestion
    """Split a :py:class:`TensorMap` into multiple :py:class:`TensorMap`.
```

**Comment by PicoCentauri** [comment]
I removed everything else then the basic operations. I have a local backup branch with everything. Once we decide to bring them in, we do not have to write everything again from scratch...

File: python/tests/tensor.py [hunk 15]
```diff
@@ -334,3 +332,31 @@ def test_components_to_properties(self):
...
+        assert equistore.add(tensor, 1) == tensor + 1
+
+    def test_sub(self, tensor):
+        assert equistore.subtract(tensor, 1) == tensor - 1
+
+    def test_mul(self, tensor):
+        assert equistore.multiply(tensor, 2) == tensor * 2
+
+    def test_matmul(self, tensor):
+        tensor = tensor
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply python/tests/tensor.py:352]: Me neither, maybe it is leftover from the rebase...

**Pull request merged**